### PR TITLE
txn: make load data transactional

### PR DESCRIPTION
**Pull request by ekexium**
### What problem does this PR solve?

Issue Number: ref #49078

The PR is currently intended for test and discussion.

### What changed and how does it work?

```
Make `LOAD DATA` statements behave normally in a transaction.
Previously the `LOAD DATA` statement commits the current transaction, start a new transaction for its own use and commit it immediately after the statement.
The PR fixes the issue and make `LOAD DATA` behave like other normal DMLs. It doesn't create txn for its own and respects the rollback or commit action of its surrounding transaction.
```

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Make `LOAD DATA` statements behave normally in a transaction.
Previously the `LOAD DATA` statement commits the current transaction, start a new transaction for its own use and commit it immediately after the statement.
The PR fixes the issue and make `LOAD DATA` behave like other normal DMLs. It doesn't create txn for its own and respects the rollback or commit action of its surrounding transaction.
```

### CODE REVIEW

**Comment by cfzjywxk** [comment]
> Experimental change to make `LOAD DATA` statements behave normally in a transaction.
@ekexium The description could be updated.

**Comment by ekexium** [comment]
> > Experimental change to make `LOAD DATA` statements behave normally in a transaction.
> 
> @ekexium The description could be updated.

Thanks. Modified.

File: pkg/server/conn.go [hunk 1]
```diff
@@ -2064,17 +1994,125 @@ func (cc *clientConn) handleStmt(ctx context.Context, stmt ast.StmtNode, warns [
...
+				// check kill flag again, let the draining loop could quit if empty packet could not be received
+				if atomic.CompareAndSwapUint32(
+					&cc.ctx.GetSessionVars().SQLKiller.Signal,
+					1,
+					0,
+				) {
+					logutil.Logger(ctx).Warn("receiving kill, stop draining data, connection may be reset")
+					return exeerrors.ErrQueryInterrupted
+				}
+				curData, err1 := cc.readPacket()
```
[the remark this answers is not in the document]
Comment (ekexium) [reply pkg/server/conn.go:2087]: Do you mean we don't need this close function at all?

[the remark this answers is not in the document]
Comment (ekexium) [reply pkg/server/conn.go:2087]: IMO the separation have some benefits, though more and some duplicated code:
Clear Roles: Keeping reading and closing separate simplifies maintenance and understanding. One reads, the other cleans up.
Error Handling: Separation makes it easier to manage errors and resource cleanup, especially if something goes wrong during loading.
Concurrency Control: Separate goroutines mean we avoid complex synchronization issues, particularly in error scenarios or interrupted connections.
What do you think?

[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/server/conn.go:2087]: IMO the easier way is that we only let readerBuilder goroutine use `readPacket`. Compared with current code, we don't need the WaitGroup. No matter LoadLocal meets error or not, we always drain the `infileReader` until EOF.

File: pkg/server/internal/testserverclient/server_client.go [hunk 2]
```diff
@@ -1008,6 +1009,173 @@ func columnsAsExpected(t *testing.T, columns []*sql.NullString, expected []strin
...
+			dbt.MustExec(fmt.Sprintf("load data local infile %q into table t", path))
+			txn, err := dbt.GetDB().Begin()
+			require.NoError(t, err)
+			rows, _ := txn.Query("select * from t")
+			cli.CheckRows(t, rows, "1")
+		},
+	)
+
+	// load file in a pessimistic transaction,
+	// should acquire locks when after its execution and before it commits.
```
Comment (cfzjywxk) [inline pkg/server/internal/testserverclient/server_client.go:1087]: Is the pessimitic locks acquired in the `handlePessimisticDML` stage? 
We may also need to add tests changing the `tidb_txn_mocd` using explicit transctions.

File: pkg/executor/load_data.go [hunk 3]
```diff
@@ -737,14 +783,12 @@ func (loadDataVarKeyType) String() string {
...
 
-// LoadDataVarKey is a variable key for load data.
-const LoadDataVarKey loadDataVarKeyType = 0
-
 var (
 	_ exec.Executor = (*LoadDataActionExec)(nil)
 )
 
 // LoadDataActionExec executes LoadDataActionStmt.
+// TODO: LoadDataActionExec and its corresponding syntax is not in use and should be deleted.
```
Comment (lance6716) [inline pkg/executor/load_data.go:791]: https://github.com/pingcap/tidb/pull/49470

File: pkg/server/internal/testserverclient/server_client.go (hunk 2 above)
Comment (ekexium) [reply pkg/server/internal/testserverclient/server_client.go:1087]: > Is the pessimitic locks acquired in the handlePessimisticDML stage?
Yes

Comment (ekexium) [reply pkg/server/internal/testserverclient/server_client.go:1087]: The Go SQL library do not let us control connections manually. I set global `tidb_txn_mode` in tests instead.

**Review by zyguan — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Comment by cfzjywxk** [comment]
@lance6716 @you06 
PTAL

**Review by lance6716 — APPROVED** [review]

**Comment by lance6716** [comment]
/hold

wait for @you06 's review

**Review by you06 — APPROVED** [review]

**Pull request merged**